src/System.Linq.Parallel/tests/PlinqModesTests.cs

```diff
@@ -11,157 +11,135 @@ namespace Test
+        }
 
-            // Verify that all queries in 'easyQueries' always run in forced mode
-            for (int i = 0; i < hardQueries.Length; i++)
+        public static IEnumerable<object[]> HardQueryData(object[] counts, object[] modes)
+        {
+            // Test doesn't apply to DOP == 1.  It verifies that work is actually
+            // happening in parallel, which won't be the case with DOP == 1.
+            if (Environment.ProcessorCount == 1) yield break;
+
+            foreach (object[] results in UnorderedSources.Ranges(counts))
             {
-                Verifier verifier = new ParVerifier();
-                hardQueries[i].Invoke(ParallelExecutionMode.ForceParallelism, verifier);
-                if (!verifier.Passed)
+                foreach (var query in HardQueries())
                 {
-                    Assert.True(false, string.Format("Hard query {0} expected to run in parallel in force-parallelism mode", i));
+                    foreach (ParallelExecutionMode mode in modes)
+                    {
+                        yield return new object[] { results[0], results[1], query, mode };
+                    }
                 }
             }
         }
 
-        #region Helper Methods / Classes
+        [Theory]
+        [OuterLoop]
+        [MemberData("EasyQueryData", new[] { 1024 }, new[] { ParallelExecutionMode.Default, ParallelExecutionMode.ForceParallelism })]
+        [MemberData("HardQueryData", new[] { 1024 }, new[] { ParallelExecutionMode.ForceParallelism })]
+        public static void WithExecutionMode(Labeled<ParallelQuery<int>> labeled, int count, Labeled<Action<ParVerifier, ParallelQuery<int>>> operation, ParallelExecutionMode mode)
+        {
+            ParVerifier verifier = new ParVerifier();
+            operation.Item(verifier, labeled.Item.WithExecutionMode(mode));
+            verifier.AssertPassed();
+        }
 
-        private static void Iterate<T>(this IEnumerable<T> e)
+        [Theory]
+        [MemberData("Ranges", (object)(new int[] { 2 }), MemberType = typeof(UnorderedSources))]
+        public static void WithExecutionMode_ArgumentException(Labeled<ParallelQuery<int>> labeled, int count)
         {
-            foreach (var x in e) { }
+            ParallelQuery<int> query = labeled.Item;
+
+            Assert.Throws<ArgumentException>(() => query.WithExecutionMode((ParallelExecutionMode)2));
         }
 
-        // A class that checks whether Verify has been called from one or multiple threads.
-        private abstract class Verifier
+        [Fact]
+        public static void WithExecutionMode_ArgumentNullException()
         {
-            internal abstract int Verify(int x);
+            Assert.Throws<ArgumentNullException>(() => ((ParallelQuery<int>)null).WithExecutionMode(ParallelExecutionMode.Default));
+        }
 
-            internal abstract bool Passed { get; }
+        private static void ForEach<T>(this IEnumerable<T> e)
```

Comment:
Kind of a strange name to me, in particular because it doesn't take an Action<T> to run "for each" element.